// zzish/admins.js
/*
 * Zzish JavaScript SDK
 * 
 * 
 *	TODO
 * 	- check if /secure for api is working correctly
 */

module.exports = Admins;

var utils = require('./utils');
var config = require('./config');
var user = require('./users');

function Admins(proxy){
	utils.decorateInstance(this, {}, proxy);
}



Admins.prototype.getUsers = function(callback) {
	var proxy = this.__proxy;
	var request = {
		url: proxy.buildUrl(this.__paths.secureusers, ''),
		method: "GET",
		data: {}
	};
	
	console.log(request.url);

	proxy.ajax( request, function (err, res) {
		if (err) {
			callback(err, res);
		} else {
			console.log('get res in sdk: ' + JSON.stringify(res));
			
			var user;

			// res.forEach(function(user) {
			// 	var temp =  new user.User(user, proxy);
			
			// 	user.push(temp);
			// });
			callback(err, res);
		}
	});
};

Admins.prototype.getRoles = function(callback) {
	var proxy = this.__proxy;
	var request = {
		url: proxy.buildUrl('secure', 'roles'),
		method: "GET",
		data: {}
	};
	
	console.log(request.url);

	proxy.ajax( request, function (err, res) {
		if (err) {
			callback(err, res);
		} else {
			console.log('get res in sdk: ' + JSON.stringify(res));
			
			var user;

			callback(err, res);
		}
	});
};

Admins.prototype.getTopics = function(callback) {
	var proxy = this.__proxy;
	var request = {
		url: proxy.buildUrl('secure', 'topics'),
		method: "GET",
		data: {}
	};
	
	console.log(request.url);

	proxy.ajax( request, function (err, res) {
		if (err) {
			callback(err, res);
		} else {
			console.log('get res in sdk: ' + JSON.stringify(res));
			callback(err, res);
		}
	});
};
