// zzish/apps.js
/*
 * Zzish JavaScript SDK
 *
 * Application Module (currently private)
 *
 */

module.exports = Apps;
var utils = require('./utils');

function Apps(proxy) {
	utils.decorateInstance(this, {}, proxy);
}

/* * * * 
Create

eg params: AppObject
note: MUST INCLUDE {owner: {id: USER_ID}} (userId)
return: AppObject
* * * */

Apps.prototype.create = function (data, callback) {
	var proxy = this.__proxy;

	var request = {
		url: this.__proxy.buildUrl(this.__paths.apps),
		data: data
	};
	this.__proxy.ajax( request, function (err, res) {
		if (err) {
			callback(err, res);
		} else {
			var app = new App(res, proxy);
			callback(err, app);
		}
	});
};

/* * * * 
Get

eg params: app_id
return: AppObject
* * * */

Apps.prototype.get = function(id, callback) {
	var proxy = this.__proxy;
	var request = {
		url: proxy.buildUrl(this.__paths.apps, id),
		method: "GET",
		data: {}
	};
	
	proxy.ajax( request, function (err, res) {
		if (err) {
			callback(err, res);
		} else {
			var app = new App(res, proxy);
			callback(err, app);
		}
	});
};

function App(data, proxy){
	utils.decorateInstance(this, data, proxy);
}
